pyro/contrib/gp/models/vgp.py

```diff
@@ -149,10 +149,10 @@ def forward(self, Xnew, full_cov=False):
         :rtype: tuple(torch.Tensor, torch.Tensor)
         """
         self._check_Xnew_shape(Xnew)
-        tmp_sample_latent = self._sample_latent
+        # avoid sampling the unnecessary latent f
         self._sample_latent = False
         f_loc, f_scale_tril = self.guide()
-        self._sample_latent = tmp_sample_latent
+        self._sample_latent = True
```

Comment:
Can you confirm that `self._sample_latent` is always `True` when entering this function? If that is not the case, consider using the save-and-restore pattern
```py
saved_sample_latent = self._sample_latent
self._sample_latent = False
f_loc, f_scale_tril = self.guide()
self._sample_latent = saved_sample_latent
```

Reply:
`self._sample_latent` should always be `True` unless we want to set it `False` to achieve something. This is the only place we change its value so `self._sample_latent` will always be True when entering this function (unless users do something like `gpmodel._sample_latent = False` (I don't know why they want to do it and even if they do, this place will set `self._sample_latent` back to its `True` value).